src/binary/timestamp.rs

```diff
@@ -0,0 +1,98 @@
+    types::timestamp::{Precision, Timestamp},
+};
+
+/// Extends [`Timestamp`] to support [Ion binary].
+///
+/// [Ion binary]: https://amzn.github.io/ion-docs/docs/binary.html#6-timestamp
+impl Timestamp {
+    const BUFFER_SIZE: usize = 16;
+
+    /// NOTE: Currently, this function always encodes with nanosecond precision.
+    pub fn encode_binary(&self) -> IonResult<SmallVec<[u8; Timestamp::BUFFER_SIZE]>> {
+        const SECONDS_PER_MINUTE: f32 = 60f32;
+        let mut buffer = [0u8; Timestamp::BUFFER_SIZE];
+        let mut writer = io::Cursor::new(&mut buffer).writer();
+
+        // Each component of the timestamp is in UTC time. Readers then apply the offset minutes
+        // to derive the localized time.
+        let utc = self.date_time;
+
+        // Ion encodes offsets in minutes while DateTime stores it in seconds.
+        let offset_seconds = self.offset.map(|o| o.utc_minus_local()).unwrap_or(0);
+        let offset_minutes = (offset_seconds as f32 / SECONDS_PER_MINUTE).round() as i64;
+        VarInt::write_i64(&mut writer, offset_minutes)?;
+
+        VarUInt::write_u64(&mut writer, utc.year() as u64)?;
+        VarUInt::write_u64(&mut writer, utc.month() as u64)?;
+
+        // So far, we've written required fields. The rest are optional!
+        if self.precision > Precision::Month {
+            VarUInt::write_u64(&mut writer, utc.day() as u64)?;
+            if self.precision > Precision::Day {
+                VarUInt::write_u64(&mut writer, utc.hour() as u64)?;
+                VarUInt::write_u64(&mut writer, utc.minute() as u64)?;
+                if self.precision > Precision::HourAndMinute {
+                    VarUInt::write_u64(&mut writer, utc.second() as u64)?;
+                }
+                if self.precision > Precision::Second {
+                    // This implementation always writes the datetime out with nanosecond precision.
+                    // The nanoseconds are an integer, which means that our exponent is always 9
+                    // and our coefficient is the number of nanoseconds.
+                    let nanoseconds = utc.nanosecond();
+                    const SCALE: i64 = -9; // "Scale" (used by BigDecimal) is -1 * exponent
+                    VarInt::write_i64(&mut writer, SCALE)?;
+                    DecodedUInt::write_u64(&mut writer, nanoseconds as u64)?;
+                }
+            }
+        }
+
+        let encoded_length = writer.get_ref().position() as usize;
+        let slice = &writer.into_inner().into_inner()[..encoded_length];
+        let encoded = SmallVec::from(slice);
+
+        Ok(encoded)
+    }
+}
+
+pub trait TimestampBinaryExt {
+    fn write_timestamp(&mut self, value: &Timestamp) -> IonResult<()>;
+}
+
+impl<W> TimestampBinaryExt for W
+where
+    W: Write,
+{
+    /// Write a timestamp. Encodes the timestamp and prepends an appropriate
+    /// type decriptor.
+    fn write_timestamp(&mut self, value: &Timestamp) -> IonResult<()> {
```

Comment:
I think returning `IonResult<usize>` would be more helpful.

Should this take `Option<&Timestamp>` to capture the null case more explicitly?

Reply:
Ok!

Reply:
> Should this take Option<&Timestamp> to capture the null case more explicitly?

`ion-rust` hasn't been using this pattern elsewhere. You'd write a `null.timestamp` with `write_null(IonType::Timestamp)`.